internal/scan/iterator.go

```diff
@@ -0,0 +1,123 @@
+package scan
+
+import (
+	"context"
+	"fmt"
+	"iter"
+	"path/filepath"
+	"regexp"
+	"strconv"
+
+	"github.com/stellar/go/support/datastore"
+)
+
+var ledgerFilenameRe = regexp.MustCompile(`^[0-9A-F]{8}--[0-9]+(?:-[0-9]+)?\.xdr\.[A-Za-z0-9._-]+$`)
+
+type LedgerFile struct {
+	key  string
+	high uint32
+	low  uint32
+}
+
+type LedgerFileIter struct {
+	DS         datastore.DataStore
+	StartAfter string // upper bound: iteration starts *after* this key
+	StopAfter  string // lower bound: stop once we reach or go below this key
+}
+
+// Next returns an iterator (iter.Seq2) that yields ledger objects in
+// descending ledger sequence order.
+//
+// File paths are listed lexicographically ascending by the datastore, but
+// their embedded ledger numbers decrease, so iteration effectively walks
+// backward in ledger order.
+//
+// StartAfter is an exclusive lower bound; StopAfter is an inclusive upper
+// bound in lexicographic order. The iterator stops when no more files remain,
+// the context is canceled, or an error occurs.
+func (it *LedgerFileIter) Next(ctx context.Context) iter.Seq2[LedgerFile, error] {
+	return func(yield func(LedgerFile, error) bool) {
+		startAfter := it.StartAfter
+
+		for {
+			if err := ctx.Err(); err != nil {
+				yield(LedgerFile{}, err)
+				return
+			}
```

Comment:
no need to check the context error here because if the context is done, the context error will be immediately returned in the `.DS.ListFilePaths()` invocation below